swift/ql/test/query-tests/Security/CWE-089/SQLite.swift

```diff
@@ -0,0 +1,133 @@
+	public func prepare(_ statement: String, _ bindings: [String: Binding?]) throws -> Statement { return Statement(self, "") }
+
+	@discardableResult public func run(_ statement: String, _ bindings: Binding?...) throws -> Statement { return Statement(self, "") }
+	@discardableResult public func run(_ statement: String, _ bindings: [Binding?]) throws -> Statement { return Statement(self, "") }
+	@discardableResult public func run(_ statement: String, _ bindings: [String: Binding?]) throws -> Statement { return Statement(self, "") }
+
+	public func scalar(_ statement: String, _ bindings: Binding?...) throws -> Binding? { return nil }
+	public func scalar(_ statement: String, _ bindings: [Binding?]) throws -> Binding? { return nil }
+	public func scalar(_ statement: String, _ bindings: [String: Binding?]) throws -> Binding? { return nil }
+}
+
+// --- tests ---
+
+func test_sqlite_swift_api(db: Connection) {
+	let localString = "user"
+	let remoteString = try! String(contentsOf: URL(string: "http://example.com/")!)
+	let remoteNumber = Int(remoteString) ?? 0
+
+	let unsafeQuery1 = remoteString
+	let unsafeQuery2 = "SELECT * FROM users WHERE username='" + remoteString + "'"
+	let unsafeQuery3 = "SELECT * FROM users WHERE username='\(remoteString)'"
+	let safeQuery1 = "SELECT * FROM users WHERE username='\(localString)'"
+	let safeQuery2 = "SELECT * FROM users WHERE username='\(remoteNumber)'"
+
+	// --- execute ---
+
+	try db.execute(unsafeQuery1) // BAD
+	try db.execute(unsafeQuery2) // BAD
+	try db.execute(unsafeQuery3) // BAD
+	try db.execute(safeQuery1) // GOOD
+	try db.execute(safeQuery2) // GOOD
+
+	// --- prepared statements ---
+
+	let varQuery = "SELECT * FROM users WHERE username=?"
+
+	let stmt1 = try db.prepare(unsafeQuery3) // BAD
+	try stmt1.run()
+
+	let stmt2 = try db.prepare(varQuery, localString) // GOOD
+	try stmt2.run()
+
+	let stmt3 = try db.prepare(varQuery, remoteString) // GOOD
+	try stmt3.run()
+
+	let stmt4 = Statement(db, localString) // GOOD
+	stmt4.run()
+
+	let stmt5 = Statement(db, remoteString) // BAD
+	stmt5.run()
+
+	// --- more variants ---
+
+	let stmt6 = try db.prepare(unsafeQuery1, "") // BAD
+	try stmt6.run()
+
+	let stmt7 = try db.prepare(unsafeQuery1, [""]) // BAD
+	try stmt7.run()
+
+	let stmt8 = try db.prepare(unsafeQuery1, ["username": ""]) // BAD
+	try stmt8.run()
+
+	db.run(unsafeQuery1, "") // BAD
+
+	db.run(unsafeQuery1, [""]) // BAD
+
+	db.run(unsafeQuery1, ["username": ""]) // BAD
+
+	db.scalar(unsafeQuery1, "") // BAD
+
+	db.scalar(unsafeQuery1, [""]) // BAD
+
+	db.scalar(unsafeQuery1, ["username": ""]) // BAD
+
+	let stmt9 = try db.prepare(varQuery) // GOOD
+	stmt9.bind(remoteString)
+	stmt9.bind([remoteString])
+	stmt9.bind(["username": remoteString])
+	try stmt9.run(remoteString)
+	try stmt9.run([remoteString])
+	try stmt9.run(["username": remoteString])
+	try stmt9.scalar(remoteString)
+	try stmt9.scalar([remoteString])
+	try stmt9.scalar(["username": remoteString])
```

Comment:
These I do not quite understand, even after looking at the SQLite.swift docs. Shouldn't these be tagged with `// GOOD`, or am I missing something?